pkg/config/secret-sync-config_test.go

```diff
@@ -14,18 +14,213 @@ limitations under the License.
 package config
 
 import (
+	"context"
+	"flag"
+	"fmt"
+	"github.com/b01901143/secret-sync-controller/pkg/client"
+	"github.com/b01901143/secret-sync-controller/tests"
+	"os"
+	"reflect"
 	"testing"
+	"time"
 )
 
+var testClient tests.ClientInterface
+
+type testOptions struct {
+	e2e        bool
+	configPath string
+	gsmProject string
+	kubeconfig string
+}
+
+var testOpts testOptions
+
+func TestMain(m *testing.M) {
+	flag.BoolVar(&testOpts.e2e, "e2e", false, "Test with actual ConfigMap.")
+	flag.StringVar(&testOpts.configPath, "config-path", "/tmp/config/syncConfig", "Path to mounted ConfigMap.")
+	flag.StringVar(&testOpts.gsmProject, "gsm-project", "project-1", "Secret MAnager prject for e2e testing.")
+	flag.StringVar(&testOpts.kubeconfig, "kubeconfig", "", "Path to kubeconfig file.")
+	flag.Parse()
+
+	if !testOpts.e2e {
+		testClient = tests.NewMockClient([]string{testOpts.gsmProject})
+	} else {
+		// prepare clients
+		k8sClientset, err := client.NewK8sClientset(testOpts.kubeconfig)
+		if err != nil {
+			fmt.Printf("Fail to create new kubernetes client: %s", err)
+			os.Exit(1)
+		}
+		secretManagerClient, err := client.NewSecretManagerClient(context.Background())
+		if err != nil {
+			fmt.Printf("Fail to create new Secret Manager client: %s", err)
+			os.Exit(1)
+		}
+
+		testClient = &tests.E2eTestClient{
+			&client.Client{
+				K8sClientset:        *k8sClientset,
+				SecretManagerClient: *secretManagerClient,
+			},
+		}
+	}
+
+	os.Exit(m.Run())
+}
+
+func TestWatchConfig(t *testing.T) {
```

Comment:
Make it explicit this isn't a unit test
```suggestion
func TestWatchConfigE2E(t *testing.T) {
```

Reply:
I guess I would either leave this be or get rid of it, rather than spend too much time improving it.  I can't think of a great way to rewrite it at the moment.

Reply:
I guess I'm getting rid of it, it's going to take a lot of time, and I still cannot come up with a goo way to test it.